src/transformStateWithClones.js

```
@@ -6,8 +6,48 @@
  *
  * @return {Object[]}
  */
+
 function transformStateWithClones(state, actions) {
-  // write code here
+  const modifiedArr = [];
+  const finalSolution = [];
+  const modifiedState = { ...state };
+
+  modifiedArr.push(modifiedState);
+
+  let lastElement = modifiedArr[modifiedArr.length - 1];
+
+  for (const prop of actions) {
+    switch (prop.type) {
+      case 'clear': {
+        lastElement = {};
+        break;
+      }
+
+      case 'addProperties': {
+        for (const key in prop.extraData) {
+          lastElement[key] = prop.extraData[key];
+        }
+        break;
+      }
+
+      case 'removeProperties': {
+        for (const key of prop.keysToRemove) {
+          delete lastElement[key];
+        }
+        break;
+      }
+
+      default: {
+        return 'Please enter valid data';
```

The `default` branch returns a string (`'Please enter valid data'`) which will change the function's return type from an array to a string when an unknown action is passed. The checklist requires a default case for error handling: "switch/case should always have default case for error handling." Consider throwing an Error or safely ignoring unknown action types instead of returning a string so the function's API remains consistent.